ncc/examples/wu.c

```diff
@@ -0,0 +1,256 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <uvm/graphics.h>
+#include <uvm/syscalls.h>
+#include <uvm/utils.h>
+
+
+size_t FRAME_WIDTH = 400;
+size_t FRAME_HEIGHT = 400;
+
+u32 frame_buffer[160000];
+
+// Current mouse pointer position
+size_t pos_x = 200;
+size_t pos_y = 200;
+
+void
+carefree_alpha_blend_plot_pixel(u32* dest, size_t dest_stride, u64 x, u64 y, u32 color, u8 alpha)
+{
+    if (!alpha) return;
+    u32* pix_ptr = dest + dest_stride * y + x;
+    if (0xFF == alpha)
+    {
+        *pix_ptr = color;
+        return;
+    }
+    u32 dest_pixel = *pix_ptr;
+    u8 unalpha = 0xFF - alpha;
+    u8 red = (((dest_pixel >> 16) & 255) * unalpha + ((color >> 16) & 255) * alpha) / 0xff;
+    u8 green = (((dest_pixel >> 8) & 255) * unalpha + ((color >> 8) & 255) * alpha) / 0xff;
+    u8 blue = ((dest_pixel & 255) * unalpha + (color & 255) * alpha) / 0xff;
+    *pix_ptr = (red << 16) | (green << 8) | blue;
+}
+
+
+// > the “error adjust,” is stored as a fixed-point fraction, in 0.16 format (that is, all bits are fractional, and the decimal point is just to the left of bit 15). An error accumulator, also in 0.16 format, is initialized to 0. Then the first pixel is drawn; no weighting is needed, because the line intersects its endpoints exactly.
+
+// LIMIT is 16 bits of 0.999... in 0.16 format
+#define LIMIT 0xffff
+
+// Draw a line using Wu's algorithm
+void draw_wu_line(u32* fb, u32 fb_width, u32 fb_height, u32 x0, u32 y0, u32 x1, u32 y1, u32 color)
+{
+    if (x1 == x0)
+    {
+        if (y1 == y0)
+        {
+            //draw_point
+        }
+        else
+        {
+            //draw_vertical_line
+        }
+    }
+    else if (y1 == y0)  // and we know x1 != x0
+    {
+        //draw_horizontal_line
+    }
```

Comment:
Is this case missing from the implementation?

Reply:
Oh right!  I'll add the special cases tonight.

Reply:
Hey, I think I should be able to get to this today.  (It's been an eventful week.)

Reply:
Nice. Please make sure to test all the edge cases :)

Reply:
I haven't forgotten about this PR, I've just been really busy.  I'll get back to it soon.

The last time I sync'd my branch to main and ran the demo there was some problem with the colors.  I haven't yet had time to revisit that and figure out what's wrong.  Hopefully it just goes away after then next pull, eh?

That reminds me, the font demo had a few spurious pixels above the first glyph.  I never figured out what that was about, and now it's moot.  Such is life, eh?

Reply:
> I haven't forgotten about this PR, I've just been really busy. I'll get back to it soon.

No stress :)

If you can find a small repro wrt color issues, open an issue and I'll take a look 👍

Reply:
Hey, I've been busy, but I took a moment to look at this today and I found out what was causing the messed up colors.  This change fixed it (`red`, `green`, and `blue` are `u8`):

``` C
-    *pix_ptr = (red << 16) | (green << 8) | blue;
+    *pix_ptr = rgb32(red, green, blue);
```